Modules/_threadmodule.c

```diff
@@ -2426,13 +2419,16 @@ _thread_set_name_impl(PyObject *module, PyObject *name_obj)
 
     const char *name = PyBytes_AS_STRING(name_encoded);
 #ifdef PYTHREAD_NAME_MAXLEN
-    // Truncate to PYTHREAD_NAME_MAXLEN bytes + the NUL byte
-    char buffer[PYTHREAD_NAME_MAXLEN + 1];
+    // Truncate to PYTHREAD_NAME_MAXLEN bytes + the NUL byte if needed
     size_t len = strlen(name);
     if (len > PYTHREAD_NAME_MAXLEN) {
-        memcpy(buffer, name, PYTHREAD_NAME_MAXLEN);
-        buffer[PYTHREAD_NAME_MAXLEN] = 0;
-        name = buffer;
+        PyObject *truncated = PyBytes_FromStringAndSize(name, PYTHREAD_NAME_MAXLEN);
+        if (truncated == NULL) {
+            Py_DECREF(name_encoded);
+            return NULL;
+        }
+        Py_SETREF(name_encoded, truncated);
```

Comment:
It can be moved before `if`. Then `Py_DECREF` will not be needed.

Reply:
I prefer the code as it is.